ext/ddtrace_profiling_native_extension/heap_recorder.c

```diff
@@ -0,0 +1,489 @@
+  object_record_free((object_record *) value);
+  return ST_DELETE;
+}
+
+void heap_recorder_free(struct heap_recorder* recorder) {
+  st_foreach(recorder->object_records, st_object_table_entry_free, 0);
+  st_free_table(recorder->object_records);
+
+  st_foreach(recorder->heap_records, st_record_table_entry_free, 0);
+  st_free_table(recorder->heap_records);
+
+  pthread_mutex_destroy(&recorder->records_mutex);
+
+  ruby_xfree(recorder->reusable_locations);
+  ruby_xfree(recorder->queued_samples);
+
+  ruby_xfree(recorder);
+}
+
+typedef struct {
+  void (*for_each_callback)(stack_iteration_data stack_data, void *extra_arg);
+  void *for_each_callback_extra_arg;
+  heap_recorder *heap_recorder;
+} internal_iteration_data;
+
+static int st_heap_records_iterate(st_data_t key, st_data_t value, st_data_t extra) {
+  heap_stack *stack = (heap_stack*) key;
+  heap_record *record = (heap_record*) value;
+  internal_iteration_data *iteration_data = (internal_iteration_data*) extra;
+
+  ddog_prof_Location *locations = iteration_data->heap_recorder->reusable_locations;
+
+  for (uint64_t i = 0; i < stack->frames_len; i++) {
+    heap_frame *frame = &stack->frames[i];
+    ddog_prof_Location *location = &locations[i];
+    location->function.name.ptr = frame->name;
+    location->function.name.len = strlen(frame->name);
+    location->function.filename.ptr = frame->filename;
+    location->function.filename.len = strlen(frame->filename);
+    location->line = frame->line;
+  }
+
+  stack_iteration_data stack_data;
+  stack_data.inuse_objects = record->inuse_objects;
+  stack_data.locations = (ddog_prof_Slice_Location) {.ptr = locations, .len = stack->frames_len};
+
+  iteration_data->for_each_callback(stack_data, iteration_data->for_each_callback_extra_arg);
+
+  return ST_CONTINUE;
+}
+
+void heap_recorder_iterate_stacks(heap_recorder *heap_recorder, void (*for_each_callback)(stack_iteration_data stack_data, void *extra_arg), void *for_each_callback_extra_arg) {
+  pthread_mutex_lock(&heap_recorder->records_mutex);
+  internal_iteration_data internal_iteration_data;
+  internal_iteration_data.for_each_callback = for_each_callback;
+  internal_iteration_data.for_each_callback_extra_arg = for_each_callback_extra_arg;
+  internal_iteration_data.heap_recorder = heap_recorder;
+  st_foreach(heap_recorder->heap_records, st_heap_records_iterate, (st_data_t) &internal_iteration_data);
+  pthread_mutex_unlock(&heap_recorder->records_mutex);
+}
+
+void commit_allocation(heap_recorder *heap_recorder, heap_stack *heap_stack, VALUE obj, unsigned int weight) {
+  heap_record *heap_record = NULL;
+  if (!st_lookup(heap_recorder->heap_records, (st_data_t) heap_stack, (st_data_t*) &heap_record)) {
+    heap_record = heap_record_init(heap_stack);
+    if (st_insert(heap_recorder->heap_records, (st_data_t) heap_stack, (st_data_t) heap_record)) {
+      rb_raise(rb_eRuntimeError, "Duplicate heap stack tracking: %p", heap_stack);
+      return;
```

Comment:
I'm curious, when would this happen?

Reply:
Ah, this is one of those `throw "Should never happen"` that in a few months actually shows up in prod :rofl: